python/KMS_KEYS_TO_NOT_DELETE.py

```diff
@@ -0,0 +1,452 @@
+
+'''
+#####################################
+##            Gherkin              ##
+#####################################
+
+Rule Name:
+	KMS_KEYS_TO_NOT_DELETE
+
+Description:
+	Check that Customer Managed keys are not scheduled for deletion. The rule is NON_COMPLAINT if the Customer Managed Keys are scheduled for deletion. This rule does not include AWS-managed and Imported Keys.
+
+Trigger:
+	Periodic
+
+Reports on:
+	AWS::KMS::Key
+
+Rule Parameters:
+	kmsKeyIds (Optional)
+	Comma-separated list of specific Customer Managed Key Ids, that are expected not to be scheduled for deletion.
+
+Scenarios:
+	Scenario 1:
+	Given: No CMKs in the Account
+	Then: Return NOT_APPLICABLE
+
+    	Scenario 2:
+    	Given: At least 1 CMK is present in the AWS Account
+    	And: Rule parameter kmsKeyIds are configured and not valid
+    	Then: Return ERROR
+
+	Scenario 3:
+	Given: At least 1 CMK is present in the AWS Account
+	And: Rule parameter kmsKeyIds are not configured
+	And: The KMS Key is not scheduled for deletion
+	Then: Return COMPLIANT
+
+	Scenario 4:
+	Given: At least 1 CMK is present in the AWS Account
+	And: Rule parameter kmsKeyIds are not configured
+	And: The KMS Key in the account is scheduled for deletion
+	Then: Return NON_COMPLIANT
+
+	Scenario 5:
+	Given: At least 1 CMK is present in the AWS Account
+	And: Rule parameter kmsKeyIds are configured and valid
+	And: The KMS Key is not one of the keys in the parameter
+	Then: Return NOT_APPLICABLE
+
+	Scenario 6:
+	Given: At least 1 CMK is present in the AWS Account
+	And: Rule parameter kmsKeyIds are configured and valid
+	And: The CMK is one of the keys in the parameter
+	And: The CMK is not scheduled for deletion
+	Then: Return COMPLIANT
+
+	Scenario 7:
+	Given: At least 1 CMK is present in the AWS Account
+	And: Rule parameter kmsKeyIds are configured and valid
+	And: The CMK is one of the keys in the parameter
+	And: The CMK is scheduled for deletion
+	Then: Return NON_COMPLIANT
+'''
+
+import json
+import sys
+import datetime
+import re
+import boto3
+import botocore
+
+try:
+    import liblogging
+except ImportError:
+    pass
+
+##############
+# Parameters #
+##############
+
+# Define the default resource to report to Config Rules
+DEFAULT_RESOURCE_TYPE = 'AWS::KMS::Key'
+
+# Set to True to get the lambda to assume the Role attached on the Config Service (useful for cross-account).
+ASSUME_ROLE_MODE = False
+
+# Other parameters (no change needed)
+CONFIG_ROLE_TIMEOUT_SECONDS = 900
+
+#############
+# Main Code #
+#############
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    evaluations = []
+    kms_client = get_client('kms', event)
+    account_keys = get_account_keys(kms_client)
+
+    if not account_keys:
+        return None
+
+    if 'kmsKeyIds' in valid_rule_parameters:
+        for key_id in valid_rule_parameters['kmsKeyIds']:
+            if key_id in account_keys:
```

Comment:
The use case when the key_id is not in the `account_keys` is not covered in the Gherkin, please add it. I suggest you to return NOT_APPLICABLE with a comment. It will allow the customer to troubleshoot.